// patternfly-react/dist/esm/components/LoginPage/components/LoginCardComponents/LoginLanguagePicker.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React from 'react';
import PropTypes from 'prop-types';
import { DropdownButton, MenuItem } from '../../../../index';
import { noop } from '../../../../common/helpers';

var LoginLanguagePicker = function (_React$Component) {
  _inherits(LoginLanguagePicker, _React$Component);

  function LoginLanguagePicker(props) {
    _classCallCheck(this, LoginLanguagePicker);

    var _this = _possibleConstructorReturn(this, (LoginLanguagePicker.__proto__ || Object.getPrototypeOf(LoginLanguagePicker)).call(this, props));

    _this.handleClick = function (e) {
      _this.props.onLanguageChange && _this.props.onLanguageChange(e);
      _this.setState({ title: e.target.text });
    };

    _this.state = {
      title: _this.props.selectedLanguage && _this.props.selectedLanguage.text || _this.props.availableLanguages && _this.props.availableLanguages[0].text
    };
    return _this;
  }

  _createClass(LoginLanguagePicker, [{
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      var selectedLanguage = nextProps.selectedLanguage,
          availableLanguages = nextProps.availableLanguages;

      var title = selectedLanguage && selectedLanguage.text || availableLanguages && availableLanguages[0].text;

      this.setState({ title: title });
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var availableLanguages = this.props.availableLanguages;

      if (!availableLanguages) {
        return null;
      }
      var menuItems = availableLanguages.map(function (language, index) {
        return React.createElement(
          MenuItem,
          {
            key: index,
            value: language.value,
            active: _this2.state.title === language.text,
            onClick: function onClick(e) {
              return _this2.handleClick(e);
            }
          },
          language.text
        );
      });

      return React.createElement(
        'div',
        { className: 'bootstrap-select btn-group ' + this.props.className },
        React.createElement(
          DropdownButton,
          { title: this.state.title, id: this.props.id },
          menuItems
        )
      );
    }
  }]);

  return LoginLanguagePicker;
}(React.Component);

LoginLanguagePicker.propTypes = {
  availableLanguages: PropTypes.array,
  selectedLanguage: PropTypes.shape({
    value: PropTypes.string,
    text: PropTypes.string
  }),
  onLanguageChange: PropTypes.func,
  className: PropTypes.string,
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
};

LoginLanguagePicker.defaultProps = {
  selectedLanguage: null,
  availableLanguages: null,
  onLanguageChange: noop,
  className: '',
  id: 'language-picker'
};

export default LoginLanguagePicker;
