// patternfly-react/dist/esm/components/LoginPage/components/LoginCardComponents/LoginCardSocialColumns.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React from 'react';
import PropTypes from 'prop-types';
import { Button, Icon } from '../../../../index';
import LoginCardSocialLink from './LoginCardSocialLink';

var LoginCardSocialColumns = function (_React$Component) {
  _inherits(LoginCardSocialColumns, _React$Component);

  function LoginCardSocialColumns() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, LoginCardSocialColumns);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = LoginCardSocialColumns.__proto__ || Object.getPrototypeOf(LoginCardSocialColumns)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      expend: false,
      width: window.innerWidth
    }, _this.updateWindowWidth = function () {
      _this.setState({
        width: window.innerWidth
      });
    }, _this.getListItems = function () {
      _this.hiddenLinks = [];
      return _this.props.links && _this.props.links.map(function (link, index) {
        if (index >= _this.props.shownButtons) {
          _this.hiddenLinks.push(link);
          return true;
        }

        return React.createElement(LoginCardSocialLink, { link: link, key: link.key || index });
      });
    }, _this.getHiddenListItems = function () {
      return _this.hiddenLinks && _this.hiddenLinks.map(function (link, index) {
        return React.createElement(LoginCardSocialLink, { link: link, key: link.key || index + _this.props.shownButtons });
      });
    }, _this.toggleExpend = function () {
      _this.setState({ expend: !_this.state.expend });
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(LoginCardSocialColumns, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      window.addEventListener('resize', this.updateWindowWidth);
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      window.removeEventListener('resize', this.updateWindowWidth);
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          links = _props.links,
          shownButtons = _props.shownButtons;

      if (!links) {
        return null;
      }
      var _state = this.state,
          expend = _state.expend,
          width = _state.width;

      var expendButton = width > 768 && links.length > shownButtons && React.createElement(
        Button,
        { bsStyle: 'link', bsClass: 'btn btn-link login-pf-social-toggle', onClick: function onClick(e) {
            return _this2.toggleExpend(e);
          } },
        expend ? 'Less' : 'More',
        ' \xA0',
        React.createElement(Icon, { name: 'angle-' + (expend ? 'up' : 'down') })
      );

      var doubleColumn = links.length > 4 ? 'login-pf-social-double-col' : '';
      var moreItems = expend || width < 768 ? this.getHiddenListItems() : null;
      return React.createElement(
        'div',
        null,
        React.createElement(
          'ul',
          { className: 'login-pf-social list-unstyled ' + doubleColumn },
          this.getListItems(),
          moreItems
        ),
        expendButton
      );
    }
  }]);

  return LoginCardSocialColumns;
}(React.Component);

LoginCardSocialColumns.propTypes = {
  links: PropTypes.array,
  shownButtons: PropTypes.number
};

LoginCardSocialColumns.defaultProps = {
  links: [],
  shownButtons: 8
};

export default LoginCardSocialColumns;
